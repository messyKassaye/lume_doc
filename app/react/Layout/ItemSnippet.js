import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';

import SafeHTML from 'app/utils/SafeHTML';
import getFieldLabel from 'app/Templates/utils/getFieldLabel';

import t from '../I18N/t';

export const ItemSnippet = ({ snippets, onSnippetClick, template }) => {
  let content;
  let source;
  if (snippets.metadata.length) {
    source = getFieldLabel(snippets.metadata[0].field, template);
    [content] = snippets.metadata[0].texts;
  } else {
    source = t('System', 'Document contents');
    content = snippets.fullText[0].text;
  }
  /* eslint-disable react/no-danger */
  const snippetElement = (
    <>
      <div className="item-snippet-source">{source}</div>
      <div onClick={onSnippetClick} className="item-snippet">
        <SafeHTML>{content}</SafeHTML>
      </div>
    </>
  );

  if (snippets.count === 1) {
    return <div className="item-snippet-wrapper">{snippetElement}</div>;
  }

  return (
    <div className="item-snippet-wrapper">
      {snippetElement}
      <div>
        <a onClick={onSnippetClick}>{t('System', 'Show more')}</a>
      </div>
    </div>
  );
};

ItemSnippet.propTypes = {
  snippets: PropTypes.shape({
    count: PropTypes.number,
    metadata: PropTypes.array,
    fullText: PropTypes.array,
  }).isRequired,
  onSnippetClick: PropTypes.func,
  doc: PropTypes.shape({
    template: PropTypes.string,
  }).isRequired,
  template: PropTypes.shape({
    get: PropTypes.func,
  }).isRequired,
};

ItemSnippet.defaultProps = {
  onSnippetClick: undefined,
};

export const mapStateToProps = (state, ownProps) => ({
  template: state.templates.find(tmpl => tmpl.get('_id') === ownProps.doc.template),
});

export default connect(mapStateToProps)(ItemSnippet);
