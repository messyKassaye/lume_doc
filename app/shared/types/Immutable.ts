import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
