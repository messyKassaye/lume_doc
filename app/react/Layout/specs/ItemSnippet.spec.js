import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';
import { ItemSnippet } from '../ItemSnippet';

describe('ItemSnippet', () => {
  let component;
  let props;

  beforeEach(() => {
    props = {
      doc: {},
      snippets: {
        count: 0,
        metadata: [
          {
            field: 'title',
            texts: ['title 1 <b>snippet</b> found', 'title 2 <b>snippet</b> found'],
          },
          {
            field: 'metadata.summary',
            texts: ['metadata 1 <b>snippet</b> found', 'metadata 2 <b>snippet</b> found'],
          },
        ],
        fullText: [
          { text: 'doc <b>snippet</b> found', page: 1 },
          { text: 'other doc <b>snippet</b> found', page: 2 },
        ],
      },
      onSnippetClick: jest.fn(),
      template: Immutable.fromJS({
        _id: 'template',
        commonProperties: [{ name: 'title', label: 'Title' }],
        properties: [
          {
            name: 'summary',
            label: 'Summary',
          },
        ],
      }),
    };
  });

  const render = () => {
    component = shallow(<ItemSnippet {...props} />);
  };

  it('should first metadata snippet if exists', () => {
    render();
    expect(component).toMatchSnapshot();
  });
  it('should first metadata snippet when there are no document content snippets', () => {
    render();
    expect(component).toMatchSnapshot();
  });
  it('should first title snippet if metadata only has title snippets', () => {
    props.snippets.metadata = [
      {
        field: 'title',
        texts: ['title 1 <b>snippet</b> found'],
      },
    ];
    render();
    expect(component).toMatchSnapshot();
  });
  it('should first metadata snippet if metadata only has metadata field snippets', () => {
    props.snippets.metadata = [
      {
        field: 'metadata.summary',
        texts: ['metadata 1 <b>snippet</b> found'],
      },
    ];
    render();
    expect(component).toMatchSnapshot();
  });
  it('should show first document snippet if there are not metadata snippets', () => {
    props.snippets.metadata = [];
    render();
    expect(component).toMatchSnapshot();
  });
  it('should call onSnippetClick when clicking on the snippet', () => {
    render();
    component.find('.item-snippet').simulate('click');
    expect(props.onSnippetClick).toHaveBeenCalled();
  });
});
