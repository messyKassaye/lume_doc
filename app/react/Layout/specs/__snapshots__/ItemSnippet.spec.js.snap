// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSnippet should first metadata snippet if exists 1`] = `
<div
  className="item-snippet-wrapper"
>
  <div
    className="item-snippet-source"
  >
    Title
  </div>
  <div
    className="item-snippet"
    onClick={[MockFunction]}
  >
    <SafeHTML>
      title 1 &lt;b&gt;snippet&lt;/b&gt; found
    </SafeHTML>
  </div>
  <div>
    <a
      onClick={[MockFunction]}
    >
      Show more
    </a>
  </div>
</div>
`;

exports[`ItemSnippet should first metadata snippet if metadata only has metadata field snippets 1`] = `
<div
  className="item-snippet-wrapper"
>
  <div
    className="item-snippet-source"
  >
    Summary
  </div>
  <div
    className="item-snippet"
    onClick={[MockFunction]}
  >
    <SafeHTML>
      metadata 1 &lt;b&gt;snippet&lt;/b&gt; found
    </SafeHTML>
  </div>
  <div>
    <a
      onClick={[MockFunction]}
    >
      Show more
    </a>
  </div>
</div>
`;

exports[`ItemSnippet should first metadata snippet when there are no document content snippets 1`] = `
<div
  className="item-snippet-wrapper"
>
  <div
    className="item-snippet-source"
  >
    Title
  </div>
  <div
    className="item-snippet"
    onClick={[MockFunction]}
  >
    <SafeHTML>
      title 1 &lt;b&gt;snippet&lt;/b&gt; found
    </SafeHTML>
  </div>
  <div>
    <a
      onClick={[MockFunction]}
    >
      Show more
    </a>
  </div>
</div>
`;

exports[`ItemSnippet should first title snippet if metadata only has title snippets 1`] = `
<div
  className="item-snippet-wrapper"
>
  <div
    className="item-snippet-source"
  >
    Title
  </div>
  <div
    className="item-snippet"
    onClick={[MockFunction]}
  >
    <SafeHTML>
      title 1 &lt;b&gt;snippet&lt;/b&gt; found
    </SafeHTML>
  </div>
  <div>
    <a
      onClick={[MockFunction]}
    >
      Show more
    </a>
  </div>
</div>
`;

exports[`ItemSnippet should show first document snippet if there are not metadata snippets 1`] = `
<div
  className="item-snippet-wrapper"
>
  <div
    className="item-snippet-source"
  >
    Document contents
  </div>
  <div
    className="item-snippet"
    onClick={[MockFunction]}
  >
    <SafeHTML>
      doc &lt;b&gt;snippet&lt;/b&gt; found
    </SafeHTML>
  </div>
  <div>
    <a
      onClick={[MockFunction]}
    >
      Show more
    </a>
  </div>
</div>
`;
